src/main/java/hudson/plugins/ec2/util/EC2GuiceModule.java

```diff
@@ -0,0 +1,23 @@
+package hudson.plugins.ec2.util;
+
+import com.google.inject.AbstractModule;
+import com.google.inject.Singleton;
+
+import hudson.Extension;
+
+@Extension
+public class EC2GuiceModule extends AbstractModule {
+    @Override
+    protected void configure() {
```

Comment:
You're going to want to bind some real things instead of configuring the class name like this. See https://github.com/google/guice/wiki/LinkedBindings for some introductory info. Also:

* https://github.com/google/guice/wiki/ProvidesMethods for the method that returns an `AmazonEC2` instance
* https://github.com/google/guice/wiki/InstanceBindings for the URL value binding.

Or, if you want to really get fancy similar to what you're doing here, see https://github.com/google/guice/wiki/JustInTimeBindings

Reply:
I'm not sure I follow entirely. I do understand the `@Provides` part, but won't that interfere with my ability to override actual implementation class?

As far as I understand it, `@Provides` does basically the same as `bind(MethodReturnType.class).to(ImplementationType.class)`. I will be able to get rid of the factory that way though.

However, the good part about having the Factory injectable, but not the AmazonEC2 implementation is that the `AmazonEC2Client` class is in the AWS SDK and has no default constructor but requires a lot of parameters. Thus, for the real implementation, I need to supply all that stuff in the `new AmazonEC2Client(.....)` but in the mock factory I can just do `Mockito.mock(AmazonEC2.class)`

Reply:
And if I move the instance creation to a `@Provides` method, that would require `Mockito.mock` in production code, right? Or am I misinterpreting something?

Is it possible to have two `@Provides` and give the test one priority when running tests?

Reply:
I may have found the answer in an old Guice issue:

https://github.com/google/guice/issues/1029

`Modules.override()` should be what I need. I'll just have a production module and a test module.

Reply:
Right, you want two different Module classes here as they essentially correspond to a configuration.